### PR TITLE
Image param func

### PR DESCRIPTION
**Pull request by psuriana**
This PR adds support for implicit conversion of ImageParam into a Func.  It also adds support for in() directives in ImageWrapper (see #1182). 

The implicit conversion into Func allows you to pass ImageParam around without having to convert the ImageParam to a Func manually. For example, if `helper` takes  a Func as argument, in the previous version, you would have to do something like this:
```
ImageParam img(Int(32), 2);
helper(lambda(x, y, img(x, y)));
```
, with the implicit conversion, you can simply do this instead:
```
ImageParam img(Int(32), 2);
helper(img);
```

The in() directives in ImageParam lets you do something like this:
```
     ImageParamImageParam img(Int(32), 2, "img");
output(x, y) = img(y, x);
output.compute_root().gpu_tile(x, y, 8, 8);
img.in().compute_at(final, Var::gpu_blocks()).unroll(_0, 2).unroll(_1, 2).gpu_threads(x, y);
 img(Int(32), 2);
     Func g;
     g(x, y) = ... img(x, y) ...
     img.in(g)
```
, which will re-write the pipeline into a pipeline like this:

 ```
     ImageParam img(Int(32), 2);
     Func img_wrap, g;
     img_wrap(x, y) = img(x, y);
     g(x, y) = ... img_wrap(x, y) ...
```

This can potentially be useful when you want to stage loads from ImageParam into some intermediate buffer, let's say the GPU shared memory:
```
ImageParam img(Int(32), 2, "img");
output(x, y) = img(y, x);
output.compute_root().gpu_tile(x, y, 8, 8);
img.in().compute_at(output, Var::gpu_blocks()).unroll(_0, 2).unroll(_1, 2).gpu_threads(x, y);
```

### CODE REVIEW

**Comment by abadams** [comment]
This diff is going to be unreadably large until we merge in(), which should be today, I think.

File: src/Func.cpp [hunk 1]
```diff
@@ -863,6 +865,108 @@ void Func::invalidate_cache() {
...
+    const auto &iter = wrappers.find(f.name());
+    if (iter == wrappers.end()) {
+        Func wrapper(name() + "_in_" + f.name());
+        wrapper(args()) = (*this)(args());
+        wrapper.func.freeze();
+        func.add_wrapper(wrapper.func, f.name());
+        return wrapper;
+    }
+
+    IntrusivePtr<FunctionContents> wfunc = iter->second;
```
Comment (steven-johnson) [inline src/Func.cpp]: Nit, but: "wfunc" is an awkward name.

File: src/Func.cpp [hunk 2]
```diff
@@ -863,6 +865,108 @@ void Func::invalidate_cache() {
...
+            << Function(wfunc).name() << "] in " << f.name() << " is not allowed\n";
+    }
+    Function wrapper(wfunc);
+    internal_assert(wrapper.frozen());
+    return Func(wrapper);
+}
+
+Func Func::in(const vector<Func>& fs) {
+    if (fs.empty()) {
+        return in(); // Make a global wrapper
```
Comment (steven-johnson) [inline src/Func.cpp]: Are we sure this is the behavior we want? (ie, should it be an error condition to pass in an empty vector?) Not clear to me that the fallback should be to a global wrapper.

Comment (abadams) [inline src/Func.cpp]: This is code from the other PR, but I agree that this should be an error. Func::in({}) would mean wrap this Func where it is used in... nothing at all. Probably indicates a bug in the calling code.

**Comment by steven-johnson** [comment]
> This diff is going to be unreadably large until we merge in(), which should be today, I think.

OK, I'll wait till then for further review.

File: src/ImageParam.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+     ImageParam im(UInt(8), 2);
+     Func f;
+     f = im*2;
+     \endcode
+     *
+     * This will define f as a two-dimensional function with value at
+     * position (x, y) equal to twice the value of the image parameter
+     * at the same location.
+     */
+    operator Expr() const {
```
Comment (abadams) [inline src/ImageParam.h]: I think this operator should be removed. The comment is wrong: f = im*2 is not legal syntax. We removed operator Expr from Func recently, and I should have removed it here too, but forgot.

File: src/ImageParam.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+/** An Image parameter to a halide pipeline. E.g., the input image. */
+class ImageParam : public OutputImageParam {
+
+    /** Func representation of the ImageParam.
+     * All call to ImageParam is equivalent to call to its intrinsic Func
+     * representation. */
+    Func func;
+
+    // Helper function to initialize the Func representation of this ImageParam
```
Comment (abadams) [inline src/ImageParam.h]: Probably want to use a doxygen-style comment here too /** ... */

File: src/OutputImageParam.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+ * promises about the output size and stride. */
+class OutputImageParam {
+protected:
+    /** A reference-counted handle on the internal parameter object */
+    Internal::Parameter param;
+
+    /** Is this an input or an output? OutputImageParam is the base class for both. */
+    Argument::Kind kind;
+
+    void add_implicit_args_if_placeholder(std::vector<Expr> &args,
```
Comment (abadams) [inline src/OutputImageParam.h]: It's possible this function can be deleted now.

**Comment by abadams** [comment]
lgtm. Some minor comments.

**Pull request merged**